ALPHA_BETA.md

```diff
@@ -0,0 +1,109 @@
+# rubot: alpha_beta
+
+The goal of the `Bot` is to take a decision tree and output
+the optimal decision.
+
+At each depth of this tree we either want to maximise or minimize
+the resulting fitness.
+
+In a competitive two player game, maximizing depths are the turns in which the bot
+is active, and during minimizing depths the opponent is able to make a decision.
+
+- `a n` := the bot is active, meaning we want to take the path which results in the highest fitness.
+    the node itself has a fitness of `n`.
+- `o n` := the opponent is active, meaning that we want to achieve the lowest possible fitness.
+
+Paths are represented as a list of indices.
+
+## Examples
+
+```
+a 0
+  - a 12 # child 0
+  - a 7  # child 1
+
+optimal action: [0], with fitness: 12
+
+a 0
+  - a 12
+    - a 6
+  - a 7
+
+optimal action: [1], with fitness: 7
+
+a 0
+  - o 12
+    - a 3
+    - a 7
+  - a 6
+
+optimal action: [1], with fitness: 6
+```
+
+## Cutoffs
+
+Unlike the brute bot, alpha beta pruning allows us to skip subtrees if
+it is provable that they always result less desirable outcome than an already computed
+path.
+
+```
+a 0
+  - a 6
+  - o 12
+    - a 3
+    - a 7
+```
+Walking through this decision tree from top to bottom, we know by visiting `[0]` that
+there is a path with the fitness `6`. While analysing `[1]`, we realise that
+our opponent could chose `[1][0]`, which would result in a fitness of `3`.
+
+As `3` is worse than `6`, we can stop looking at `[1]` entirely, without having seen `[1][1]`.
+
+It is also possible to apply this in the other direction:
+
+```
+a 0
+ - o 0
+   - a 6
+   - a 0
+     - a 7
+     - a 3
+```
+
+While analysing `[0]`, the opponent first looks at `[0][0]` which ends up with a fitness of `6`.
+Once they see `[0][1][0]` with a fitness of `7`, they known that we would always be able to choose
+a path at `[0][1]` with a fitness `> 6`. This means that they are also able to skip `[0][1][1]` entirely.
+
+- `alpha`: the worst guaranteed fitness.
+- `beta`: the currently best value, which cannot be denied by the opponent.
+- `alpha cutoff`: skips paths where the opponent could choose an option which is worse than `alpha`.
+- `beta cutoff`: while minimizing, skips paths where we could choose an option better than `beta`.
+
```

Comment:
So the first of two previous example is `alpha cutoff` and the second is `beta cutoff`, right? It would be also instructive to track alpha and beta values as we analyse the tree.